sfaira/data/store/carts/multi.py

```diff
@@ -59,20 +59,19 @@ def iterator(self, repeat: int = 1, shuffle_buffer: int = 0):
 
         if self.intercalated:
             while keep_repeating:
-                ratios = self.ratios.copy()
-                # Document which generators are still yielding batches:
-                yielding = np.ones((ratios.shape[0],)) == 1.
+                iterator_frequencies = self.iterator_frequencies.copy()
                 iterators = [v.iterator(repeat=1, shuffle_buffer=shuffle_buffer) for v in self.carts.values()]
-                while np.any(yielding):
-                    # Loop over one iterator length adjusted cycle of emissions.
-                    for i, (gi, n) in enumerate(zip(iterators, ratios)):
-                        n_rounded = int(n) + np.random.binomial(n=1, p=n - int(n))
-                        for _ in range(n_rounded):
-                            try:
-                                x = next(gi)
-                                yield x
-                            except StopIteration:
-                                yielding[i] = False
+                while len(iterators) > 0:
+                    # Sample iterator with frequencies so that in expectation, the frequency of samples from each
+                    # iterator is uniform over an epoch.
+                    itertor_idx = np.random.multinomial(n=1, pvals=iterator_frequencies, partial=())
```

Comment:
Can we add the single organism per batch case here as well? That we yield batch_size samples here of a single organism. Which then get concatenated into a single batch by the pytorch dataloader

Reply:
But have to make sure this gets consumed right by the pytorch dataloader